lib/controllers/session_controller.dart

```diff
@@ -88,82 +93,104 @@ class SessionController extends ValueNotifier<AgoraSettings> {
             uid: value.connectionData!.uid,
           );
           await value.engine?.renewToken(token);
+          agoraEventHandlers?.tokenPrivilegeWillExpire!(token);
         },
         remoteVideoStateChanged: (uid, state, reason, elapsed) {
           if (state == VideoRemoteState.Stopped) {
             updateUserVideo(uid: uid, videoDisabled: true);
-          } else if (state == VideoRemoteState.Decoding && reason == VideoRemoteStateReason.RemoteUnmuted) {
+          } else if (state == VideoRemoteState.Decoding &&
+              reason == VideoRemoteStateReason.RemoteUnmuted) {
             updateUserVideo(uid: uid, videoDisabled: false);
           }
+          agoraEventHandlers?.remoteVideoStateChanged!(
+              uid, state, reason, elapsed);
         },
         remoteAudioStateChanged: (uid, state, reason, elapsed) {
           if (state == AudioRemoteState.Stopped) {
             updateUserAudio(uid: uid, muted: true);
-          } else if (state == AudioRemoteState.Decoding && reason == AudioRemoteStateReason.RemoteUnmuted) {
+          } else if (state == AudioRemoteState.Decoding &&
+              reason == AudioRemoteStateReason.RemoteUnmuted) {
             updateUserAudio(uid: uid, muted: false);
           }
+          agoraEventHandlers?.remoteAudioStateChanged!(
+              uid, state, reason, elapsed);
         },
         localAudioStateChanged: (state, error) {
           if (state == AudioLocalState.Stopped) {
             updateUserAudio(uid: value.localUid, muted: true);
           } else if (state == AudioLocalState.Recording) {
             updateUserAudio(uid: value.localUid, muted: false);
           }
+          agoraEventHandlers?.localAudioStateChanged!(state, error);
```

Comment:
Do a null check before this and all other places, since localAudioStateChanged can be null here